packages/vite/src/node/plugin.ts

```diff
@@ -120,67 +314,80 @@ export interface Plugin<A = any> extends RollupPlugin<A> {
-   * list of modules affected by the file change, and the dev server instance.
-   *
-   * - The hook can return a filtered list of modules to narrow down the update.
-   *   e.g. for a Vue SFC, we can narrow down the part to update by comparing
-   *   the descriptors.
-   *
-   * - The hook can also return an empty array and then perform custom updates
-   *   by sending a custom hmr payload via server.hot.send().
-   *
-   * - If the hook doesn't return a value, the hmr update will be performed as
-   *   normal.
+   * @deprecated
+   * Compat support, ctx.modules is a backward compatible ModuleNode array
+   * with the mixed client and ssr moduleGraph. Use hotUpdate instead
    */
   handleHotUpdate?: ObjectHook<
     (
       this: void,
       ctx: HmrContext,
     ) => Array<ModuleNode> | void | Promise<Array<ModuleNode> | void>
   >
-
-  /**
-   * extend hooks with ssr flag
-   */
-  resolveId?: ObjectHook<
-    (
-      this: PluginContext,
-      source: string,
-      importer: string | undefined,
-      options: {
-        attributes: Record<string, string>
-        custom?: CustomPluginOptions
-        ssr?: boolean
-        /**
-         * @internal
-         */
-        scan?: boolean
-        isEntry: boolean
-      },
-    ) => Promise<ResolveIdResult> | ResolveIdResult
-  >
-  load?: ObjectHook<
-    (
-      this: PluginContext,
-      id: string,
-      options?: { ssr?: boolean },
-    ) => Promise<LoadResult> | LoadResult
-  >
-  transform?: ObjectHook<
-    (
-      this: TransformPluginContext,
-      code: string,
-      id: string,
-      options?: { ssr?: boolean },
-    ) => Promise<TransformResult> | TransformResult
-  >
 }
 
 export type HookHandler<T> = T extends ObjectHook<infer H> ? H : T
 
 export type PluginWithRequiredHook<K extends keyof Plugin> = Plugin & {
   [P in K]: NonNullable<Plugin[P]>
 }
+
+export type BoundedPluginConstructor = (
+  Environment: PluginEnvironment,
+) => BoundedPluginOption
+
+export type MaybeBoundedPlugin = BoundedPlugin | false | null | undefined
+
+export type BoundedPluginOption =
+  | MaybeBoundedPlugin
+  | BoundedPluginOption[]
+  | Promise<MaybeBoundedPlugin | BoundedPluginOption[]>
+
+export type MaybePlugin = Plugin | false | null | undefined
+
+export type PluginOption =
+  | MaybePlugin
+  | PluginOption[]
+  | Promise<MaybePlugin | PluginOption[]>
+
+export async function resolveBoundedPlugins(
+  environment: PluginEnvironment,
+): Promise<BoundedPlugin[]> {
+  const resolvedPlugins: BoundedPlugin[] = []
+  for (const plugin of environment.config.plugins) {
+    if (plugin.create) {
+      const boundedPlugin = await plugin.create(environment)
+      if (boundedPlugin) {
+        const flatPlugins = await asyncFlattenBoundedPlugin(
+          environment,
+          boundedPlugin,
+        )
+        resolvedPlugins.push(...flatPlugins)
+      }
+    } else {
+      resolvedPlugins.push(plugin)
+    }
```

Comment:
I noticed that when using `create`, the other hooks won't be used. I had something like this:

```ts
{
  name: "my-unocss",

  // wanted to run transform for everything,
  transform(code, id) {
    ...
  },

  create(environment) {
    // other hooks to split the behavior
    ...
  }
}
```

Is this limitation intended? I think it's fine to restrict the use case if there's a reason, but noting it in jsdoc would help.

Reply:
Ha, I told @antfu yesterday that this was already implemented yesterday! It should work as you expect after https://github.com/vitejs/vite/pull/16471/commits/12d467f642a0d71fcc718d036d20dc4b6efda549, released as vite@6.0.0-alpha.8. The new plugins returned by `create` are inserted in the pipeline after the plugin that creates them. You can still use `config`, `transform`, etc in it.

Reply:
Thanks for the update! I'm now checking alpha8 https://github.com/hi-ogawa/vite-environment-examples/pull/62 and I'm wondering if `enforce` is going to work inside `create`. Probably I need to use `order` instead?

```ts
// this logs in this order
//   [create-order-pre]
//   [outer]
//   [create-enforce-pre]

{
  name: "outer",

  transform(code, id) {
    console.log("[outer]", id)
  },

  create(environment) {
     return [
        {
          name: "create-order-pre",
          transform: {
            order: "pre",
            handler(code, id) {
              console.log("[create-order-pre]", { id })
            },
          },
        },
        {
          name: "create-enforce-pre",
          enforce: "pre",
          transform(code, id) {
            console.log("[create-enforce-pre]", { id })
          }
        }
     ]
  }
}
```

Reply:
Yes, you can check the types for `BoundedPlugin`, it only adds `hotUpdate` to Rollup plugin (and `this.environment` in all hooks, eventually). All other hooks and extensions (`enforce`, `apply`) are only used in regular plugins right now.

We could add support for `enforce` and `apply` in bounded plugins though. This would be helpful if you want to use a regular plugin inside of create 🤔

There will be a few kind of plugins from now on:

- Shared only Plugins (using `config`, `configureServer`, `create`, `this.environment`)
- Per Environment only Plugins (using `environment` inside `create`)
- Plugins that you works if you add them to the shared pipeline or inside create (support for json for example)

If possible, plugin authors should create their plugins to fall in the last category so it is more flexible for users.

I'll check if we can add an extra re-ordering and apply step after resolving bounded plugins 👍🏼

Reply:
Yeah, unless it's adding more confusion or complication internally, I think supporting `enforce` and `apply` would be nice.

Also regarding `BoundedPlugin`, since this type wasn't exposed, I was pushing to `Plugin` array inside `create` and I didn't notice the type difference. Can you expose the type?
I actually tried to extract the type from something like `ReturnType<NonNullable<Plugin["create"]>>`, but could't figure it out.

```ts
import { type Plugin } from "vite";

const sharedPlugin: Plugin = {
  name: "...",
  create() {
    // would've noticed if this were BoundedPlugin[]
    const plugins: Plugin[] = [];

    plugins.push({
      name: "...",
      enforce: "pre", 
    });
    
    return plugins;
  }
}
```

Reply:
I'll expose the type 👍🏼 (I think we may need a better name for it though... I'm having a hard time naming so many new things)

About `enforce` and `apply`, I don't know what is best yet. I can find good reasons to go either way. Probably something to discuss with others.

Reply:
For reference, we ended up allowing `enforce`. For now, I think it is better to do a direct conditional instead of using `apply`, but maybe we can also respect it to make more of current ecosystem plugins usable as per-environment plugins without modifications.